File: BackTracking/Combination_Sum_II.py
# Problem Link :- https://leetcode.com/problems/combination-sum-ii/

"""Given a collection of candidate numbers (candidates) and a target number (target), 
find all unique combinations in candidates where the candidate numbers sums to target.
Each number in candidates may only be used once in the combination."""

class Solution(object):
    def combinationSum2(self, candidates, target):
        ans = []
        candidates.sort()
        return self.helper(candidates,target,ans,[],0)
        
    def helper(self,nums,target,ans,path,ind):
        if target < 0:
            return 
        elif target == 0:
            if path not in ans:
                ans.append(path)
                return 
        for i in range(ind,len(nums)):
            self.helper(nums,target-nums[i],ans,path+[nums[i]],i+1)
        return ans
        

File: BackTracking/CombinationSumII.cpp
//Problem link:- https://leetcode.com/problems/combination-sum-ii/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    void helper(vector<int>& candidates,int target,int index,int sm,vector<int>path, vector<vector<int>>& result){
        if (sm == target){
            result.push_back(path);
            return;
        }
        if (sm > target)
            return;
        for(int i=index;i<candidates.size();i++){
            if (i == index || candidates[i] != candidates[i-1]){
                path.push_back(candidates[i]);
                helper(candidates,target,i+1,sm+candidates[i],path,result);
                path.pop_back();
            }
        }
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        if (candidates.size() == 0)
            return result;
        vector<int> path;
        sort(candidates.begin(),candidates.end());
        helper(candidates,target,0,0,path,result);
        return result;
    }
};